File: python/array_search/meta_binary_search_production.py
from dataclasses import dataclass


@dataclass(frozen=True)
class MetaBinarySearchResult:
    found: bool 
    index: int 
    insertion_point: int 


def _highest_power_of_two_below(length):
    bit = 1 

    while bit <= length // 2:
        bit *= 2

    return bit 


def meta_binary_lower_bound(values, target, *, key=None):
    key_fn = key or (lambda value: value)
    length = len(values)

    if length == 0:
        return 0

    position = -1 
    bit = _highest_power_of_two_below(length)

    while bit > 0: 
        next_index = position + bit 

        if next_index < length and key_fn(values[next_index]) < target:
            position = next_index

        bit //= 2

    return position + 1 


def meta_binary_search_first(values, target, *, key=None):
    insertion_point = meta_binary_lower_bound(values, target, key=key)
    key_fn = key or (lambda value: value)
    found = insertion_point < len(values) and key_fn(values[insertion_point]) == target

    return MetaBinarySearchResult(
        found,
        insertion_point if found else -1,
        insertion_point
    )

File: cpp/array_search/meta_binary_search_production.cpp
#include <functional>
#include <iterator>

template <typename Iterator>
struct MetaBinarySearchResult {
    bool found;
    Iterator position; 
    Iterator insertion_point;
};

template <typename Difference>
Difference highest_power_of_two_below(Difference length) {
    Difference bit = 1;

    while (bit <= length / 2) {
        bit *= 2;
    }

    return bit;
}

template <typename RandomIt, typename T, typename Compare = std::less<>>
RandomIt meta_binary_lower_bound(RandomIt first, RandomIt last, const T& target, Compare compare = Compare{}) {
    using Difference = typename std::iterator_traits<RandomIt>::difference_type;
    Difference length = last - first;

    if (length <= 0) {
        return first;
    }

    Difference position = -1;
    Difference bit = highest_power_of_two_below(length);

    while (bit > 0) {
        Difference next = position + bit;

        if (next < length && compare(first[next], target)) {
            position = next;
        }

        bit /= 2;
    }

    return first + position + 1;
}

template <typename RandomIt, typename T, typename Compare = std::less<>>
MetaBinarySearchResult<RandomIt> meta_binary_search_first(RandomIt first, RandomIt last, const T& target, Compare compare = Compare{}) {
    RandomIt position = meta_binary_lower_bound(first, last, target, compare);
    bool found = position != last && !compare(*position, target) && !compare(target, *position);
    return {found, found ? position : last, position};
}
